### PR TITLE
Add mimalloc for shared memory

### PR DESCRIPTION
**Pull request by dashanji**
What do these changes do?
-------------------------

The pr is to add mimalloc for shared memory in vineyard. Once the CI pass, I will add some benchmark results.

Related #671

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -342,6 +342,25 @@ macro(find_jemalloc)
...
+        INSTALL_COMMAND ${CMAKE_MAKE_PROGRAM}
+    )
+    add_library(mimalloc STATIC IMPORTED)
+    add_dependencies(mimalloc libmimalloc)
+    set_target_properties(mimalloc PROPERTIES IMPORTED_LOCATION
+                                                  ${PROJECT_SOURCE_DIR}/thirdparty/mimalloc/out/release/libmimalloc.a)
+    set_target_properties(mimalloc PROPERTIES INTERFACE_INCLUDE_DIRECTORIES
+                                              ${PROJECT_SOURCE_DIR}/thirdparty)
+    endif()
+endmacro(find_mimalloc)
```
Comment (sighingnow) [inline CMakeLists.txt:362]: To include dependencies with cmake support, please use `add_subdirectory_static` macro.

File: docs/notes/objects.rst [hunk 2]
```diff
@@ -319,13 +319,13 @@ listening to ipc_socket ``/var/run/vineyard.sock2``.
...
     >>> # persist the object to make it visible to form the global object
-    >>> client.persist(id1)
+    >>> client1.persist(id1)
 
     >>> # build the second array in the second vineyard instance
     >>> client2 = vineyard.connect('/var/run/vineyard.sock2')
     >>> id2 = client2.put(np.ones(4))
     >>> # persist the object to make it visible to form the global object
-    >>> client.persist(id2)
+    >>> client2.persist(id2)
```
Comment (sighingnow) [inline docs/notes/objects.rst:328]: Nice catch. Thanks!

**Review by sighingnow — CHANGES_REQUESTED** [review]
LGTM.

File: CMakeLists.txt (hunk 1 above)
Comment (dashanji) [reply CMakeLists.txt:362]: Got it.

File: src/server/memory/mimalloc.cc [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  // the addr must be 64MB aligned(required by mimalloc)
+  // this will cause some loss of memory, and the maximum does not exceed 64MB
+  void* addr = reinterpret_cast<void*>(
+      (reinterpret_cast<uintptr_t>(space) + MIMALLOC_SEGMENT_ALIGNED_SIZE - 1) &
+      ~(MIMALLOC_SEGMENT_ALIGNED_SIZE - 1));
+  size_t real_size = size - ((size_t) addr - (size_t) space);
+
+  MmapRecord& record = mmap_records[addr];
+  record.fd = fd;
+  record.size = real_size;
```
Comment (sighingnow) [inline src/server/memory/mimalloc.cc:48]:
```suggestion
  MmapRecord& record = mmap_records[space];
  record.fd = fd;
  record.size = size;
```

**Review by sighingnow — APPROVED** [review]

File: src/server/memory/mimalloc.cc (hunk 3 above)
Comment (dashanji) [reply src/server/memory/mimalloc.cc]: Copy that, thanks for correcting!

**Comment by sighingnow** [comment]
Merged. Thanks!

**Pull request merged**